apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+// webhook, but there are many cases that will require asynchronous
+// signaling via the GatewayStatus block.
+type GatewaySpec struct {
+	// GatewayClassName used for this Gateway. This is the name of a
+	// GatewayClass resource.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	GatewayClassName string `json:"gatewayClassName"`
+
+	// Listeners associated with this Gateway. Listeners define
+	// logical endpoints that are bound on this Gateway's addresses.
+	// At least one Listener MUST be specified.
+	//
+	// An implementation MAY group Listeners by Port and then collapse each
+	// group of Listeners into a single Listener if the implementation
+	// determines that the Listeners in the group are "compatible". An
+	// implementation MAY also group together and collapse compatible
+	// Listeners belonging to different Gateways.
+	//
+	// For example, an implementation might consider Listeners to be
+	// compatible with each other if all of the following conditions are
+	// met:
+	//
+	// 1. Either each Listener within the group specifies the "HTTP"
+	//    Protocol or each Listener within the group specifies either
+	//    the "HTTPS" or "TLS" Protocol.
+	//
+	// 2. Each Listener within the group specifies a Hostname that is unique
+	//    within the group.
+	//
+	// 3. As a special case, one Listener within a group may omit Hostname,
+	//    in which case this Listener matches when no other Listener
+	//    matches.
+	//
+	// If the implementation does collapse compatible Listeners, the
+	// hostname provided in the incoming client request MUST be
+	// matched to a Listener to find the correct set of Routes.
+	// The incoming hostname MUST be matched using the Hostname
+	// field for each Listener in order of most to least specific.
+	// That is, exact matches must be processed before wildcard
+	// matches.
+	//
+	// If this field specifies multiple Listeners that have the same
+	// Port value but are not compatible, the implementation must raise
+	// a "Conflicted" condition in the Listener status.
+	//
+	// Support: Core
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=64
+	Listeners []Listener `json:"listeners"`
+
+	// Addresses requested for this gateway. This is optional and
+	// behavior can depend on the GatewayClass. If a value is set
+	// in the spec and the requested address is invalid, the
+	// GatewayClass MUST indicate this in the associated entry in
+	// GatewayStatus.Addresses.
+	//
+	// If no Addresses are specified, the GatewayClass may
+	// schedule the Gateway in an implementation-defined manner,
+	// assigning an appropriate set of Addresses.
+	//
+	// The GatewayClass MUST bind all Listeners to every
+	// GatewayAddress that it assigns to the Gateway.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Addresses []GatewayAddress `json:"addresses,omitempty"`
```

Comment:
I am failing to understand the difference between Gateway's `.spec.Addresses` vs Listener's `.spec.hostname` Can you provide a use-case?

Reply:
I hadn't really thought of the overlap here, but it's clear we need to document the distinction here. The `Addresses` field is primarily used to represent addresses the entire Gateway is bound to, usually IP addresses. Some cloud load balancers choose to expose a hostname instead of an IP address, which means we need to support that kind of input as well.

Hostnames on listeners are primarily about request matching, and primarily useful when more than one listener exists. To tie all of this together, it's likely that `.listener.hostname` would be used to distinguish between foo.com and bar.com listeners, but both would be part of a Gateway listening on a 1.2.3.4 address.

Reply:
Yeah we have the same concept/problem for Service type=LoadBalancer. I think this is primarily to support ELBs, but there's likely other use cases. I'll try to clarify the use cases here without mentioning specific products.

Reply:
Maybe a clearer way to describe Addresses is something like:

The Addresses field represents a *request* for the address(es) on the "outside of the Gateway", that traffic bound for this Gateway will use. This could be the IP address or hostname of an external load balancer or other networking infrastructure, or some other address that traffic will be sent to.

The `.listener.hostname` field is used to route traffic that has already arrived at the Gateway to the correct in-cluster destination.

Reply:
Will be fixed by #839